DIL ACM/read_int.cpp

```diff
@@ -0,0 +1,99 @@
+// info about me at the bottom
+#include "read_int.h"
+#include <cctype>
+using std::isdigit;
+
+int base_of(const char *token) {
+
+  if (token[0] == '\0') { // error
+    return 0;
+  }
+
+  if (token[0] == '0') {
+
+    if (token[1] == '\0') { // token is "0"?
+      return 10;
+
+    } else if (token[1] == 'x') { // dexahecimal
+      return 16;
+
+    } else if (token[1] == 'b') { // binary
+      return 2;
+
+    } else { // octal
+      return 8;
+    }
+  } else {
+    if (isdigit(token[0]))
+      return 10; // decimal
+  }
+
+  return 0; // error
+}
+
+int read_int(char c, const int base) {
+
+  int number = -1;
+  // an invalid value that should change
+
+  if (base == 16) { // base 16
+
+    number = read_int(c, 10);
+    if (number != -1)
+      return number;
+
+    if ('a' <= c && c <= 'f')
+      return c - 'a' + 10;
+
+    if ('A' <= c && c <= 'F')
+      return c - 'A' + 10;
+
+    return -1;
+  } 
+  // any other base
+  number = c - '0';
+  if (number < 0 || number >= base)
+    return -1;
+
+  return number;
+}
+
+int read_int(const char *token, int *number) {
+
+  *number = 0;
+
+  int base = base_of(token);
+  if (!base)
+    return 0;
+  // moving after the "0b.." or "0x..." ("0.." is fine)
+  if (base == 2 || base == 16)
+    token += 2;
+
+  int digit;
+
+  // the pointer 'token' is pushed to the end
+  while (*token != '\0') {
+    digit = read_int(*token, base);
+    if (digit == -1)
+      return 0; // invalid character as digit
+
+    *number *= base;
+    *number += digit;
+    if (*number < 0)
+      return 0; // overflow
+
+    ++token;
+  }
```

Comment:
Please, don't use `*number` everywhere. This is bad for multiple reasons, one of which is readability.
You could remedy this by doing e.g.:
```cpp
int num;
...  // do whatever you do in *number with num
*number = num;
```

Reply:
Is "#define num *number" a good idea? For readability, too, because "int num" could mean making a temporary variable.

Reply:
Oh no, it's a very bad idea. :) We could have a very long discussion about the preprocessor and its usages but for the sake of brevity (and not boring people), I'll try to give you some useful thoughts (if you actually care about this discussion, you may send me PM in Slack):

To start off, the preprocessor and C are 2 languages that don't communicate very well. The one language does not know about the other. Hence, the preprocessor should be used with great care.

Now, In this specific case, using this idiom introduces a couple of bad things. First, if you write `num` anywhere in your code, to the reader _it seems_ like a variable, but it's not. That's why you usually see preprocessor entities having very distinctive names (e.g. all caps). To make clear that it's some preprocessor magic.
Even if you did give it a very distinctive name, like `NUM`, now you have introduced other bad things. First, usually preprocessor entities are about constant things (e.g. `#define NUM 10`). So, the reader will probably assume that this is a constant of some kind. Then second, you have hidden the fact that you're using a pointer under the hood. You _don't_ want to hide this fact from the reader.
Third, think that someone reads this code:
```cpp
int *number = ...;
num = 2;
```
or whatever, `number` could be a function param as in our example, e.g.:
```cpp
void some_function(int *number) {
  .. // some code
  num = 2;
}
```
Basically, what I'm trying to say is that the user sees `number` at one point and then some usage of `num`. By looking at it, those seem unrelated, however, they are very related. :P 
This is hidden from the user, again making this code very hard to read and predict.

In any case, it would always be better and more readable to write `*number` instead of replacing it textually (because that's essentially what you do with this `#define` - a search and replace for `*number` with `num`). Not to mention that the `*number` might be used somewhere else in the code which is even worse, as you may not intend to replace it there.
That's why you'll see people `#undef` stuff when they don't need them to avoid such things.

Essentially, what you want is to just _not_ use a pointer. Hence, you introduce a temporary _local_ variable, do whatever you want and you just write its value to the pointer once in the end. As I said:
```cpp
int num = ...;
.. // blah blah do things with num that you would do with *number
*number = num;  // don't forget to write the value in the end.
```
That's better for you, the reader and the compiler.

> For readability, too, because "int num" could mean making a temporary variable.

Why is that bad for readability?